feat: add v8 snapshot to monorepo (#23566)
cypress-io/cypress: tooling/v8-snapshot/README.md

  - [Generating the Snapshot Script](#generating-the-snapshot-script)
  - [Snapshot Doctor: Steps to Optimize Included Modules](#snapshot-doctor-steps-to-optimize-included-modules)
  - [Strict vs. Non-Strict Mode](#strict-vs-non-strict-mode)
  - [Result of Snapshot Doctor](#result-of-snapshot-doctor)
- [Env Vars](#env-vars)

## Features

`@tooling/v8-snapshot` aids in preparing a bundle of modules of an application and/or its dependencies to
allow those modules to be snapshotted via the `mksnapshot` tool. This snapshot can then be embedded
into the Cypress Electron application.

## Snapshot Doctor

The [snapshot-doctor][snapshot-doctor] is responsible for finding the best possible
combination of fully initialized (_healthy_) modules, modules included as definitions only
(_deferred_) and modules which code cannot be modified (_norewrite_).

Please see [docs on the `SnapshotDoctor`][snapshot-doctor-class] for more info about those.

### Snapshot Creation

When creating a snapshot the bundled code provided to it is executed and whatever ends up being
in the heap at the end is then embedded into the snapshot which is loaded whenever the app
starts up.

> Once the snapshot is loaded we can retrieve fully instantiated modules from it or instantiated

Reviewer: ```suggestion
Once the snapshot is loaded we can retrieve fully instantiated modules from it or instantiate
```

them by invoking embedded functions which when called produce the `module.exports`.

However since the environment is different when generating the snapshot and not everything is
snapshottable, certain requirements need to be respected.

### Requirements

When creating a snapshot via `mksnapshot` certain requirements need to be respected:

> - cannot `require` and Node.js core modules like `fs`

Reviewer: ```suggestion
- cannot `require` any Node.js core modules like `fs`
```

The doctor starts with an empty `healState` which means it optimistically assumes that all
modules can be included in the snapshot fully initialized.

NOTE: that the healState can be derived from metadata collected during previous doctor runs,
but here we assume the simplified case.

The doctor then produces the initial snapshot script and starts by verifying the leaf modules
which are modules that have no imports of other user land modules.

Using that same bundle it produces different snapshot scripts, each making another module to be
verified be the entry point. This is parallelized via workers, i.e. a bundle will run as many
verifiers as the machine has CPUs.

Each produced snapshot script is executed inside a Node.js VM via the
[snapshot-verifier][snashot-verifier]. Any errors are observed, processed into warnings and the
necessary consequence taken.
The possible consequences affect the module we verified in the following manner:

- Defer: we need to _defer_ the module in order to prevent it from loading
- NoRewrite: we should not _rewrite_ the module as it results in invalid code
- None: no consequence, i.e. a light weight warning for informative purposes only

> Once we did this for all leaves the doctor finds all modules that only depend on those and

Reviewer: ```suggestion
Once we have done this for all leaves, the doctor finds all modules that only depend on those and
```

repeats the process.  However the bundle that is generated takes the current (somewhat less

> optimistic _heal state_ into account and rewrites the code of the dependents to _defer_ loading

Reviewer: ```suggestion
optimistic) _heal state_ into account and rewrites the code of the dependents to _defer_ loading
```

[makeAndInstallSnapshot]:https://github.com/cypress-io/cypress/blob/bbabd12c9aae3d3eef0cb077d3e8628a53eea623/tooling/v8-snapshot/src/snapshot-generator.ts#L669

[blueprint-config]:https://github.com/cypress-io/cypress/blob/bbabd12c9aae3d3eef0cb077d3e8628a53eea623/tooling/v8-snapshot/src/blueprint.ts#L51
[globals-strict-code]:https://github.com/cypress-io/cypress/blob/bbabd12c9aae3d3eef0cb077d3e8628a53eea623/tooling/v8-snapshot/src/blueprint/globals-strict.js

[generation-opts]:https://github.com/cypress-io/cypress/blob/bbabd12c9aae3d3eef0cb077d3e8628a53eea623/tooling/v8-snapshot/src/snapshot-generator.ts#L112
[resolver-map]:https://github.com/cypress-io/cypress/blob/bbabd12c9aae3d3eef0cb077d3e8628a53eea623/tooling/v8-snapshot/src/snapshot-generator.ts#L126
[snapshot-verifier]:https://github.com/cypress-io/cypress/blob/bbabd12c9aae3d3eef0cb077d3e8628a53eea623/tooling/v8-snapshot/src/snapshot-verifier.ts#L10

[snapshot-doctor]:https://github.com/cypress-io/cypress/blob/bbabd12c9aae3d3eef0cb077d3e8628a53eea623/tooling/v8-snapshot/src/doctor/snapshot-doctor.ts

> [snapshot-doctor-class]:https://github.com/cypress-io/cypress/blob/bbabd12c9aae3d3eef0cb077d3e8628a53eea623/tooling/v8-snapshot/src/doctor/snapshot-doctor.ts#L261

Reviewer: nit: duplicate of `[doctor-class]` link above

Author: Good catch. Updated here: [`3accc37` (#23566)](https://github.com/cypress-io/cypress/pull/23566/commits/3accc371181a876d646b3c3cf5c35dd60440be3b)
